File: GSAlgorithm/C/src/gs.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include "gs.h"


bool isFree(Person* person) {
    if (person == NULL) {
        fprintf(stderr, "provided person must be non-null");
        return false;

    };
    return person->engagedTo == NULL;
};

bool hasProposedToEveryone(Person* person, int numPeople) {
    if (numPeople < 0) {
        fprintf(stderr, "Provided number of people must be non-negative");
        return true;
    };

    if (person == NULL) {
        fprintf(stderr, "person must be non-null");
        return true;
    };

    return person->numProposed >= numPeople;
};

Person* highestRankedHasNotProposedTo(Person* person, int sizeOfPreferenceList) {
    if (person == NULL) {
        fprintf(stderr, "person must be non-null");
        return NULL;
    };

    if (sizeOfPreferenceList <= 0) {
        fprintf(stderr, "preference list must be non-empty");
        return NULL;
    };

    if (person->numProposed >= sizeOfPreferenceList) {
        return NULL;
    };

    return person->preferenceList[person->numProposed];
};

bool prefersTo(Person* person, Person* candidateOne, Person* candidateTwo) {
    if (person == NULL) {
        fprintf(stderr, "person must be non-null");
        return false;
    };

    if (candidateOne == NULL || candidateTwo == NULL) {
        fprintf(stderr, "candidates must be non-null");
        return false;
    };

    if (person->preferenceList == NULL) {
        fprintf(stderr, "preference list must be non-null");
        return false;
    };


    int tracker = 0;
    Person* currentPerson = person->preferenceList[0];

    while(true) {
        if (currentPerson == candidateOne) {
            return true;
        };

        if (currentPerson == candidateTwo || currentPerson == NULL) {
            return false;
        };

        tracker += 1;
        currentPerson = person->preferenceList[tracker];
    };

    
};

Person*** GS(Person** men, Person** women, int size) {
    if (men == NULL || women == NULL) {
        fprintf(stderr, "list of men and women must not be null");
        return NULL;
    };

    if (size <=0 ) {
        fprintf(stderr, "list of men and women must be non-empty");
        return NULL;
    };

    Person*** matchList = (Person***)malloc(size * sizeof(Person**));
    if (matchList == NULL) {
        return NULL;
    };

    for (int i = 0; i < size; i++) {
        matchList[i] = (Person**)malloc(2 * sizeof(Person*));
        if (matchList[i] == NULL) {
            return NULL;
        };
    };

    bool aManIsFreeAndHasNotProposedToEveryone = true;
    while (aManIsFreeAndHasNotProposedToEveryone) {
        aManIsFreeAndHasNotProposedToEveryone = false;
        Person* currentMan = NULL;
        for (int i = 0; i < size; i++) {
            currentMan = men[i];
            if (currentMan == NULL) {
                fprintf(stderr, "lists must conform to provided size");
                return NULL;
            };
            if (isFree(currentMan) && !hasProposedToEveryone(currentMan, size)) {
                aManIsFreeAndHasNotProposedToEveryone = true;
                Person* nextWoman = highestRankedHasNotProposedTo(currentMan, size);
                if (isFree(nextWoman)) {
                    currentMan->engagedTo = nextWoman;
                    nextWoman->engagedTo = currentMan;
                    currentMan->numProposed+=1;
                } else if (prefersTo(nextWoman, nextWoman->engagedTo, currentMan)) {
                    currentMan->numProposed+=1;
                } else {
                    Person* manToFree = nextWoman->engagedTo;
                    currentMan->engagedTo = nextWoman;
                    nextWoman->engagedTo = currentMan;
                    manToFree->engagedTo = NULL;
                    currentMan->numProposed+=1;
                };
            };
            
        };
    };

    for (int i = 0; i < size; i++) {
        matchList[i][0] = men[i];
        matchList[i][1] = men[i]->engagedTo;
    };

    return matchList;

};

File: GSAlgorithm/Py/gs.py
class Person:
    def __init__(self, engagement: Person | None = None, preferenceList: list[Person] | None = None, numProposals: int = 0):
        self.engagement = engagement
        self.preferenceList = preferenceList if preferenceList else []
        self.numProposals = numProposals

    def engage(self, engagement: Person | None) -> None:
        self.engagement = engagement

    def addProposal(self) -> None:
        self.numProposals += 1

    def getSizeOfPreferenceList(self) -> int:
        if self.preferenceList is None:
            raise ValueError("preference list must be non-null")
        return len(self.preferenceList)

    def addPreferences(self, preference: Person) -> None:
        assert self.preferenceList is not None
        self.preferenceList.append(preference)

    def isFree(self) -> bool:
        return self.engagement is None

    def hasProposedToEveryone(self) -> bool:
        return self.numProposals >= self.getSizeOfPreferenceList()

    def highestRankedHasNotProposedTo(self) -> Person | None:
        if self.preferenceList is None or self.hasProposedToEveryone():
            return None

        return self.preferenceList[self.numProposals]

    def prefersTo(self, candidateOne: Person | None, candidateTwo: Person | None) -> bool:
        if candidateOne is None or candidateTwo is None:
            raise ValueError("candidates must be non-null")

        if self.preferenceList is None:
            raise ValueError("preference list must be non-null")

        for person in self.preferenceList:
            if person == candidateOne:
                return True

            if person == candidateTwo:
                return False

        raise ValueError("none of the provided candidates are present in the preference list")


def GS(men: list[Person], women: list[Person]) -> list[tuple[Person, Person | None]]:

    if men is None or women is None:
        raise ValueError("list of men and women must not be null")

    if len(men) != len(women):
        raise ValueError("number of men must be equal to number of women")

    if len(men) == 0:
        raise ValueError("must provide non-empty lists of people for a valid matching")

    matchList = []
    aManIsFreeAndHasNotProposedToEveryone = True

    while aManIsFreeAndHasNotProposedToEveryone:
        aManIsFreeAndHasNotProposedToEveryone = False
        for man in men:
            if man.isFree() and not man.hasProposedToEveryone():
                aManIsFreeAndHasNotProposedToEveryone = True
                nextWoman = man.highestRankedHasNotProposedTo()
                assert nextWoman is not None
                if nextWoman.isFree():
                    man.engage(nextWoman)
                    nextWoman.engage(man)
                    man.addProposal()
                    continue

                currentEngagement = nextWoman.engagement
                assert currentEngagement is not None
                if nextWoman.prefersTo(currentEngagement, man):
                    man.addProposal()
                else:
                    manToFree = currentEngagement
                    man.engage(nextWoman)
                    nextWoman.engage(man)
                    manToFree.engage(None)
                    man.addProposal()

    for man in men:
        matchList.append((man, man.engagement))

    return matchList